c10/util/order_preserving_flat_hash_map.h

```diff
@@ -699,27 +711,51 @@ class sherwood_v3_table : private EntryAlloc, private Hasher, private Equal
     {
         if (begin_it == end_it)
             return { begin_it.current };
-        for (EntryPointer it = begin_it.current, end = end_it.current; it != end; ++it)
-        {
-            if (it->has_value())
-            {
-                it->destroy_value();
-                --num_elements;
-            }
+
+        // whenever an entry is removed and there are other entries with the same
+        // hash, the other entries must get moved to their desired position.
+        // any reference to a moved entry is invalidated.
+        // here, we iterate through the range and collect all of the entries
+        // that we need to remove. if we invalidate an entry, we update it
+        // to its new location.
+
+        ska::flat_hash_map<EntryPointer,size_t> ptr_to_order;
+        std::vector<EntryPointer> erase_list;
+
+        size_t num_to_remove = 0;
+        for (EntryPointer it = begin_it.current, end = end_it.current; it != end;) {
+          ptr_to_order[it] = num_to_remove;
```

Comment:
Edit: realize, you only need to check pointer invalidation for `curr->next` and for `end`... update incoming...